### PR TITLE
[android] Release gradle builds zipalign and sign

### PR DESCRIPTION
**Pull request by nicknovitski**
We can use the gradle plugin to handle this instead of tracking a bunch of intermediate files

Although speaking of which, before I merge this, I need to run the PR and be sure we're adding the right path to artifacts.

### CODE REVIEW

**Comment by sjchmiela** [comment]
**TL;DR** The `STRING_TOO_LARGE` error is caused by `stripe-android` dependency, I'll try to upgrade it and see if it will make the error go away.

---

The `error: string too large to encode using UTF-8 written instead as 'STRING_TOO_LARGE'.` showed already when we upgraded Gradle in https://github.com/expo/expo/pull/2716. I've disregarded this message then, given that the command in the [SO answer](https://stackoverflow.com/a/52816708/1123156) you've linked show no such results (as did searches for `STRING_TOO_LARGE` in `build` dir). However, I've decided to dig into why it shows.

It turns out that the error message is being logged in `aapt2/StringPool.cpp#380`:

```cpp
    // Make sure the lengths to be encoded do not exceed the maximum length that
    // can be encoded using chars
    if ((((size_t)encoded.size()) > EncodeLengthMax<char>())
        || (((size_t)utf16_length) > EncodeLengthMax<char>())) {

      diag->Error(DiagMessage() << "string too large to encode using UTF-8 "
          << "written instead as '" << kStringTooLarge << "'");

      EncodeString(kStringTooLarge, utf8, out, diag);
      return false;
    }
```

Unfortunately the message doesn't include the offending string so we can't debug where it is easily. I've tried compiling AAPT2 on my own, but I couldn't find any comprehensive information on how to do this on macOS. 🤷‍♂️ 



Since it's AAPT2 which throws the error and it is a command line tool, I've come up with an idea to intercept the options passed to the executable and see which options set makes AAPT prints the message. AAPT2 used by my Gradle instance was located in `~/.gradle/caches/transforms-1/files-1.1/aapt2-3.2.1-4818971-osx.jar/40836961de362a0c71ee9b4da17faffe/aapt2-3.2.1-4818971-osx`. It soon turned out it won't be that straightforward, as Gradle runs several AAPT2 daemon instances and passes commands through standard input, so a simple `echo $@` won't work.

A simple Ruby proxy using [`Open3.popen3`](https://ruby-doc.org/stdlib-1.9.3/libdoc/open3/rdoc/Open3.html) also didn't work (dunno why). Fortunately, with
```bash
#!/bin/bash
FILE=`mktemp`
cat - | tee $FILE.in | /Users/sjchmiela/.gradle/caches/transforms-1/files-1.1/aapt2-3.2.1-4818971-osx.jar/40836961de362a0c71ee9b4da17faffe/aapt2-3.2.1-4818971-osx/aapt2-orig m | tee $FILE.out
```

I was able to create 5 sets of in/out files in general following pattern
```
c
--legacy
-o
/Users/sjchmiela/Applications/expo/android/app/build/intermediates/res/merged/prodKernel/release
/Users/sjchmiela/Applications/expo/android/app/src/main/res/drawable-hdpi/ic_arrow_back_white_36dp.png

......
......

c
--legacy
-o
/Users/sjchmiela/Applications/expo/android/app/build/intermediates/res/merged/prodKernel/release
/Users/sjchmiela/.gradle/caches/transforms-1/files-1.1/appcompat-v7-27.1.1.aar/a382a892e7800ea2f09e54acbfd4d2d4/res/drawable-xhdpi-v4/abc_ic_menu_copy_mtrl_am_alpha.png


quit
```

As described by [AAPT2 documentation](https://developer.android.com/studio/command-line/aapt2#compile), these commands prepare and compile provided resources (`mergeResources`).

By passing each one of those 5 IN files (`aapt2 m < infile`) I was able to find the one that printed the error message. It contained one other command passed to AAPT2 daemon at the end — [linking one](https://developer.android.com/studio/command-line/aapt2#link):
```
l
-I
/Users/sjchmiela/Library/Android/sdk/platforms/android-27/android.jar
--manifest
/Users/sjchmiela/Applications/expo/android/app/build/intermediates/merged_manifests/prodKernelRelease/processProdKernelReleaseManifest/merged/AndroidManifest.xml
-o
/Users/sjchmiela/Applications/expo/android/app/build/intermediates/processed_res/prodKernelRelease/processProdKernelReleaseResources/out/resources-prodKernelRelease.ap_
-R
/Users/sjchmiela/Applications/expo/android/app/build/intermediates/res/merged/prodKernel/release/drawable-xxxhdpi-v4_ic_error_amex.png.flat
-R
/Users/sjchmiela/Applications/expo/android/app/build/intermediates/res/merged/prodKernel/release/layout_design_layout_tab_icon.xml.flat

... many
... other
... -R options pointing to compiled resources

--auto-add-overlay
--java
/Users/sjchmiela/Applications/expo/android/app/build/generated/not_namespaced_r_class_sources/prodKernelRelease/processProdKernelReleaseResources/r
--proguard
/Users/sjchmiela/Applications/expo/android/app/build/intermediates/proguard-rules/prodKernel/release/aapt_rules.txt
--custom-package
host.exp.exponent
-0
apk
--output-text-symbols
/Users/sjchmiela/Applications/expo/android/app/build/intermediates/symbols/prodKernel/release/R.txt
--no-version-vectors
```

By extracting this single method to a separate infile I was able to consistently reproduce the error message.

Thank heaven AAPT2 also has a `-v` (verbose) option implemented. Running the command above with `-v` produced a verbose output containing:

```
note: linking /Users/sjchmiela/.gradle/caches/transforms-1/files-1.1/stripe-android-6.0.0.aar/733c8f3878eec2a0d496b6c3cb250594/res/drawable-anydpi-v21/ic_jcb.xml (host.exp.exponent:drawable/ic_jcb).
res/drawable-anydpi-v21/ic_jcb.xml: note: writing to archive (keep_raw_values=false).
note: writing res/drawable-anydpi-v21/ic_jcb.xml to archive.
note: linking /Users/sjchmiela/.gradle/caches/transforms-1/files-1.1/stripe-android-6.0.0.aar/733c8f3878eec2a0d496b6c3cb250594/res/drawable-anydpi-v21/ic_jcb_template_32.xml (host.exp.exponent:drawable/ic_jcb_template_32).
res/drawable-anydpi-v21/ic_jcb_template_32.xml: note: writing to archive (keep_raw_values=false).
error: string too large to encode using UTF-8 written instead as 'STRING_TOO_LARGE'.
note: writing res/drawable-anydpi-v21/ic_jcb_template_32.xml to archive.
```

It looked like `/Users/sjchmiela/.gradle/caches/transforms-1/files-1.1/stripe-android-6.0.0.aar/733c8f3878eec2a0d496b6c3cb250594/res/drawable-anydpi-v21/ic_jcb_template_32.xml` may be the offending file we were looking for. And it was! 🎉  The JCB logo SVG path written inside was 73882 characters long (more than the limit, 64k).

zrzut ekranu 2018-11-19 o 13 54 49

I'll try upgrading `stripe-android` to some more recent version containing at least this commit https://github.com/stripe/stripe-android/pull/547/files and see if the error goes away.

**Comment by nicknovitski** [comment]
👏 😲

File: android/app/build.gradle [hunk 1]
```diff
@@ -58,12 +69,7 @@ android {
     release {
       minifyEnabled true
       proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
-      zipAlignEnabled true
```
Comment (esamelson) [inline android/app/build.gradle]: what happened to this step?

**Review by esamelson — COMMENTED** [review]
i think this will probably cause errors if trying to make a release build locally, but that's fine, it already required some manual work to do that.

File: android/app/build.gradle (hunk 1 above)
Comment (nicknovitski) [reply android/app/build.gradle]: [The docs indicate signed apps are also zipaligned](https://developer.android.com/studio/publish/app-signing#gradle-sign)
> Now, when you build your app from the command line by invoking a Gradle task, Gradle signs your app (and runs zipalign) for you.

**Comment by nicknovitski** [comment]
I was thinking about that.  We could move to committing the encrypted release key, but I'm not crazy about that approach.  I know that fastlane has a system for distributing signing keys across teams for iOS, which many people use with success, but I don't see any similar feature for android. 😒

**Review by esamelson — APPROVED** [review]
I agree that committing the encrypted keystore is not worth it. We hardly ever need to make release builds locally, and I think the manual steps that are required to do so are worth the tradeoff of having fewer places where the keystore could potentially be exposed. So I'm fine with landing this as is.
